packages/gatsby/src/schema/extensions/index.js

```diff
@@ -0,0 +1,210 @@
+const {
+  GraphQLBoolean,
+  GraphQLNonNull,
+  GraphQLDirective,
+  GraphQLString,
+  DirectiveLocation,
+  defaultFieldResolver,
+} = require(`graphql`)
+const { GraphQLJSON } = require(`graphql-compose`)
+const report = require(`gatsby-cli/lib/reporter`)
+
+const { link, fileByPath } = require(`../resolvers`)
+const { getDateResolver } = require(`../types/date`)
+
+// Reserved for internal use
+const internalExtensionNames = [
+  `addDefaultResolvers`,
+  `createdFrom`,
+  `directives`,
+  `infer`,
+  `plugin`,
+]
+
+const typeExtensions = {
+  infer: {
+    description: `Infer field types from field values.`,
+    args: {
+      noDefaultResolvers: {
+        type: GraphQLBoolean,
+        description: `Don't add default resolvers to defined fields.`,
+        deprecationReason: `noDefaultResolvers is deprecated, annotate individual fields.`,
+      },
+    },
+  },
+  dontInfer: {
+    description: `Do not infer field types from field values.`,
+    args: {
+      noDefaultResolvers: {
+        type: GraphQLBoolean,
+        description: `Don't add default resolvers to defined fields.`,
+        deprecationReason: `noDefaultResolvers is deprecated, annotate individual fields.`,
+      },
+    },
+  },
+}
+
+const fieldExtensions = {
+  add: {
+    description: `Generic directive to add field extension.`,
+    args: {
+      extension: {
+        type: new GraphQLNonNull(GraphQLString),
+      },
+      options: {
+        type: GraphQLJSON,
+      },
+    },
+  },
+
+  addResolver: {
+    description: `Add a resolver specified by "type" to field.`,
+    args: {
```

Comment:
We have a bit of a weird discrepancy, where directives do type checking for arguments, but extensions applied directly won't. Not sure if we should solve and if we should then how. One way would be to manually use args to validate.

Reply:
Yeah good point! Hmm yes probably need to manually do this for extensions